### PR TITLE
Central state management within core (on a per framework basis)

### PR DESCRIPTION
**Pull request by Sn3b**
Sorry for my git noobness...

Original message:
Hi Nathan, and others if any :)

While looking into Effects and as a learning exercise, I have come across this example which I thought was quite nice, and for my own project I wanted to use this structure in combination with your seed.

I kind of like the result I must say, and I'd like your feedback if you have any time to look into it. I only created the PR so you can see the differences, not sure how I could do it differently, I'm a bit of noob when it comes to Git...

Feel free to reject or delete the PR otherwise.

Thanks!

Seb :)

### CODE REVIEW

**Comment by NathanWalker** [comment]
@Sn3b wow! excellent work 👍  Really appreciate this. I've been working on a large project based on this seed myself and we use ngrx-devtools there (I've been wanting to bring ngrx-devtools to the seed as well so this is great).

Regarding ngrx style/structure, I like it and am aware of those practices. I was on the fence about here for awhile since it does introduce some complexity and can be hard to follow at first with a simple out of box setup like this. Once your project grows, it really begins to help so I'm a fan.

I would definitely be open to merge this unless anyone else chimes in with an objection (I'll leave open for couple days)?

In meantime, If you could fix the tests would be great. Looks like mostly semantic issues: https://travis-ci.org/NathanWalker/angular-seed-advanced/jobs/176451314

Also, please keep heads up over next couple days as I will merge in parent changes for Angular 2.2.0 and {N} 2.4.0; if you could keep this PR up to date when those changes go in so when/if I merge this PR in a couple days it will be all ready to go. Again, great work, thank you.

**Comment by Sn3b** [comment]
Whoops, sorry I hadn't run the tests... I think it's all fixed now.

As you said, it brings a little complexity indeed, but as my project grows it actually adds some sense to have it this way. Glad you like it, honestly when I woke up I was afraid to read something like "No man! What were you thinking about! This is all wrong!" :P

Just a couple of points I'd like your feedback on:
- What do you think of having this app.state.ts within the core framework? I sort of liked it because as I said I'm adding extra external frameworks plus my own, and I didn't like composing my state within my own framework, I thought core was where it belongs.
- As you might have noticed, now the frameworks define their own state (at least the ones that need one) within their own state folder, the result is that the core state can just reference them easily and compose the main state, so when you look at the state in devtools you have a sort of separation per framework, does it make sense?
- Finally, as I said I just copy pasted code from this [example](https://github.com/ngrx/example-app), so I added a couple of extra dependencies, ngrx-store-freeze and deep-freeze (for dev purposes only). Deep-freeze isn't really a direct dependency however, I guess it's a dependency for ngrx-store-freeze, so it doesn't appear in the project.json file and is required in the config to resolve the path. I find it sort of confusing that a dependency needs fixing in the config but doesn't actually appear in the project.json file, what do you think? Should I add it to project.json just to remove some confusion?

**Comment by NathanWalker** [comment]
@Sn3b Move `app.state` potentially to a new framework and `NgModule` on it's own or into the root of `src/client`. `core` should be considered the bottom layer of the stack, in other words, it should not have dependency on anything above it. As you have here, `app.state` defines state for `i18n` and `sample`, both of which should be considered _optional_ frameworks if you will. Meaning you don't want `i18n` weaved into a framework layer at the base like `core` is.

You could view the out of box framework setup like this:
angular-seed-advanced-concept

Suggesting a new `AppStateModule` and `state` framework folder whose only responsibility would be define the beginning (initial) app state (assuming some state could be lazy loaded from other modules), setup `StoreModule` and run effects.

**Comment by Sn3b** [comment]
Sounds good. Since you already called the i18n framework just i18n, why not call this new framework ngrx? And the module AppStateModule?

And what about the category constant that both i18n and sample modules need to have access to? Would you move that too or staying in core is ok?

**Comment by NathanWalker** [comment]
i18n and sample modules could define their own state and actions (with category constant), then AppStateModule would simply import those and bring it all together. 

ngrx works as the folder name, let's go with that for now.

**Comment by NathanWalker** [comment]
@Sn3b thank you for staying on top of this PR with master change!
I am excited to merge this in if you could bring it up to date 1 last time to master, I will merge this next. Your recent changes are looking good 👍

**Comment by Sn3b** [comment]
Done :)

**Comment by NathanWalker** [comment]
Thank you @Sn3b great work!
